app/src/main/kotlin/com/intive/picover/images/viewmodel/ImagesViewModel.kt

```diff
@@ -0,0 +1,47 @@
+package com.intive.picover.images.viewmodel
+
+import android.net.Uri
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.intive.picover.images.repository.ImagesRepository
+import dagger.hilt.android.lifecycle.HiltViewModel
+import javax.inject.Inject
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+@HiltViewModel
+class ImagesViewModel @Inject constructor(private val repository: ImagesRepository) : ViewModel() {
+
+	private val _uiState = MutableStateFlow<LatestFirebaseStorageState>(LatestFirebaseStorageState.Loading)
+	val uiState: StateFlow<LatestFirebaseStorageState> = _uiState
+
+	init {
+		viewModelScope.launch(Dispatchers.IO) {
+			fetchPics().collect()
+		}
+	}
+
+	private suspend fun fetchPics(): Flow<List<Uri>> = callbackFlow<List<Uri>> {
```

Comment:
is it flow needed? in the repo you used `suspendCancellableCoroutine` which is designed for single-shot callback API so you will have always only one emit, or am i wrong?

Reply:
The best answer to that, that I found is:
`Flow is more flexible with regard to execution context. With Flow, you can use Kotlin coroutines to explicitly define the execution context by specifying a CoroutineDispatcher (e.g., Dispatchers.Main, Dispatchers.IO, or Dispatchers.Default). This allows you to control where and how the code should be executed, whether it's on the main thread, a background thread, or a custom thread pool. ` I want to use call to the repo on IO thread, therefore, the Flow seems to be the right choice. Furthermore, if I am not wrong Flow is recommended to be used with coroutines.

Reply:
And to be completely frank, I just want to learn Flow with coroutines :)

Reply:
you can specify dispatcher using `withContext()`, the flow as the documentation says `An asynchronous data **stream** that **sequentially** emits values and completes normally or with an exception.` You emit only once so that's why the flow is redundant. If you would change downloading images only once for observing firebase storage (what would be very nice, but I don't know it's possible, I know it's possible for firebase real time database but newer use firebase storage so I don't know) then the flow would be perfect choice.